Update @@toStringTag ES6 compliance

tcare opened the pull request:
Our current implementation of @@toStringTag is based on an old spec
revision from 2015. This change updates the implementation to conform to
the current spec.

The current behavior for handling legacy and toStringTag feature switching
is fragile, and I decided to leave the majority of it in and rename
ToStringTagHelper to ToStringTagHelperOld. The legacy code relies on the
helper, and will call it even when the @@toStringTag feature is disabled.

Other changes of note:
- Slightly tweaked the ordering of ToStringHelper so we can take advantage
  of the existing HostDispatch and TypedArray Debugger handling behavior
  while still retaining legacy codepaths.
- TypedArray toString handler spec update is also guarded by a flag.
- Overhauled the tests for @@toStringTag, adding a lot more type coverage.

When the feature is turned on by default, we can remove the legacy
codepaths for simplicity.

tcare commented:
@Microsoft/chakra-es please review

dilijev commented on test/es6/toStringTag.js:
```diff
@@ -5,299 +5,221 @@
-                }
+        name: "Proxy forwarding behavior",
+        body: function () {
+            for (var { tag, instance } of [...explicitTagBuiltIns]) {
```
nit: you should be able to use just `of explicitTagBuiltIns` in this case, right?

tcare replied:
Yes, thanks. Used to be more there. :)

dilijev commented on test/es6/toStringTag.js:
```diff
@@ -308,7 +230,7 @@ var tests = [
             assert.areEqual("[object before\0after]", o.toString(), "ToString implementation handles embedded null characters in @@toStringTag property value");
         }
-    },
+    }
```
nit: keeping the trailing comma lets future additions not have an extra line of diff

akroshg commented:
toStringTag.js test case is disabled in the rlexe.xml. Has that changed?

suwc commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -357,7 +357,127 @@ namespace Js
+
+    // This is the previous implementation of @@toStringTag, which is used in some cases even when the feature is disabled.
+    // Leaving this here for now due to the fragility.
+    // TODO(tcare): Remove when @@toStringTag is enabled by default.
```
>Remove when @@toStringTag is enabled by default. [](start = 20, length = 48)

Is @@toStringTag going to be enabled through experimental before enabled by default?

tcare replied:
The previous implementation already is I believe - it shows up in Kangax.

suwc replied:
Was it enabled recently? Because my 7/8/16 clone is not showing ES6ToStringTag enabled through FLAGPR_REGOVR_EXP in ConfigFlagsList.h

---
In reply to: [71397520](https://github.com/Microsoft/ChakraCore/pull/1288#discussion_r71397520) [](ancestors = 71397520)

tcare commented:
@akroshg nice catch - yes it should be re-enabled.

akroshg commented on test/es6/toStringTag.js:
```diff
@@ -5,299 +5,221 @@
-                a[Symbol.toStringTag];
-            } catch(e) {
-                assert.areEqual("[object ???]", Object.prototype.toString.call(a), "JS exceptions should have tag ???");
+                Object.prototype.toString.call(instance);
```
nit: why not change this test to assert.throws()  instead of having the try..catch and a variable?

akroshg replied:
what is an 'instance' over here?

tcare replied:
Copied it from the old one.. and it should be tagThrows.

suwc commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -357,7 +357,127 @@ namespace Js
+        {
+            tag = JavascriptOperators::GetProperty(thisArgAsObject, PropertyIds::_symbolToStringTag, scriptContext); // Let tag be the result of Get(O, @@toStringTag).
+        }
+        catch (JavascriptExceptionObject*)
```
>catch (JavascriptExceptionObject*) [](start = 8, length = 34)

Is this path exercised in unit tests?

tcare replied:
As Akrosh pointed out elsewhere, this is not spec compliant. I'll remove it.

akroshg commented on test/es6/toStringTag.js:
```diff
@@ -5,299 +5,221 @@
-            assert.isFalse(o.enumerable,   "WeakSet @@toStringTag is not enumerable");
-            assert.isTrue(o.configurable,  "WeakSet @@toStringTag is configurable");
+            function checkBuiltInPrototype(builtInName, prototype) {
+                WScript.Echo("Testing " + builtInName);
```
why are we printing this?

tcare replied:
This is left over from debugging. I found it helped to diagnose some issues where the assert wouldn't fire and show the message, especially when we are testing multiple types of built ins. I can either leave it in and make it depend on WScript.Arguments[0] != "summary" or just take it out completely.

akroshg commented on lib/Runtime/Library/TypedArray.cpp:
```diff
@@ -1404,13 +1405,25 @@ namespace Js
-            JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NeedTypedArray);
+            if (scriptContext->GetConfig()->IsES6ToStringTagEnabled())
+            {
+                return library->GetUndefinedDisplayString();
```
>GetUndefinedDisplayString [](start = 32, length = 25)

should we return GetUndefined() here instead of GetUndfinedDisplayString ? the spec says to return undefined.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -357,7 +357,127 @@ namespace Js
+        // 15. Let tag be ? Get(O, @@toStringTag).
+        try
+        {
+            tag = JavascriptOperators::GetProperty(thisArgAsObject, PropertyIds::_symbolToStringTag, scriptContext); // Let tag be the result of Get(O, @@toStringTag).
```
question mark in the 15 clause says ReturnIfAbrupt. So if the GetProperty throws an exception shouldn't we let it go instead of handling it here?

tcare replied:
Yes you are correct. I need to update the test case.

akroshg replied:
Have perf run with toStringTag enabled and see if we are okay.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -357,7 +357,127 @@ namespace Js
+
+            // 9. Else if O has an [[ErrorData]] internal slot, let builtinTag be "Error".
+        case TypeIds_Error:
+            builtInTag = library->CreateStringFromCppLiteral(_u("[object Error]"));
```
You can calll to GetErrorDisplayString.

tcare replied:
It looks like that isn't used anywhere. Maybe I can remove it from JavascriptLibrary.

GetObjectDisplayString is similar. After this change, shouldn't it be argued that anything using this should have used EntryToString? We have a use in ToLocaleString on a VariantDate.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +655,83 @@ namespace Js
+            DynamicObject* remoteObject = hostDispatchObject->GetRemoteObject();
+            if (!remoteObject)
+            {
+                Var result;
```
initialize this var.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +655,83 @@ namespace Js
         TypeId type = JavascriptOperators::GetTypeId(thisArg);
         JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        if (type == TypeIds_HostDispatch)
```
I don't understand this. Why are we handling this (I don;t see this in the spec). Shouldn't this be [object Object] ?

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +655,83 @@ namespace Js
+        }
+
+        // Dispatch to new @@toStringTag implemtation.
+        if (scriptContext->GetConfig()->IsES6ToStringTagEnabled()) {
```
nit: brace on new line.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +655,83 @@ namespace Js
+
+        // Dispatch to new @@toStringTag implemtation.
+        if (scriptContext->GetConfig()->IsES6ToStringTagEnabled()) {
+            if (type >= TypeIds_TypedArrayMin && type <= TypeIds_TypedArrayMax && !scriptContext->GetThreadContext()->IsScriptActive())
```
when the script will not be active? I understand this will be exercise from the script only.

tcare replied:
Spoke with Akrosh offline - we hit this in some debugger test cases.

tcare commented:
If the perf is okay, I'm tempted just to rip out the legacy code now and turn this on by default. Let me know what you think.

dilijev commented:
@tcare I'm all in favor of reducing complexity. This means we're turning it on by default and removing the ability to turn it off, right? (Is there usually a process or official decision on this sort of thing?) Has a perf run already been done?

tcare commented:
I'm in the process of doing the perf at the moment. If it does have an impact then Suwei's symbol inline cache that was used for isConcatSpreadable will be easily adapted to this use.

I'll discuss enabling it by default with the team in tomorrow's standup.

suwc commented:
I am all for ripping out "ToStringTagOld()". As to turning this on by default or not, I would ask the lead.

---
In reply to: [233802690](https://github.com/Microsoft/ChakraCore/pull/1288#issuecomment-233802690) [](ancestors = 233802690)

tcare commented:
Thanks for the review feedback. I have addressed most issues.

Some preliminary perf data:
35% slower in microbenchmark doing 1M iterations of Object.prototype.toString on all builtin types. This is expected due to the property lookup in the new spec.
Octane tests of interest are code-load and typescript. Perf is flat on my noisy machine, full perf results pending.

Not going to pursue perf at this time - however Suwei's @@isConcatSpreadable cache would easily be adapted to @@toStringTag as well.

We also decided that we will rip out the legacy code and enable this by default. This will be the next commit.

akroshg commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +647,84 @@ namespace Js
+                Js::CallInfo info(Js::CallFlags_Value, 1);
+                Js::Arguments args(info, values);
+                values[0] = thisArg;
+                if (hostDispatchObject->InvokeBuiltInOperationRemotely(EntryToString, args, &result))
```
I wonder if there is a way to test this scenario. @Yongqu do you know?

Yongqu replied:
we should be able to use the LoadScriptFile(, "crossthread") to test test. Did we add EntryToString operation in the remote support in es5 timeframe?

akroshg commented:
Overall looks good to me.

Yongqu commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -357,7 +357,119 @@ namespace Js
+        // If we don't have a tag or it's not a string, use the 'built in tag'.
+        if (isArray)
+        {
+            // 5. If isArray is true, let builtinTag be "Array".
```
we don't need to worry about proxy override toString tag here?

tcare replied:
Good catch. This should be setting the default tag, not returning it.

tcare replied:
This should be handled by the GetProperty above. Is that what you meant?

tcare commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +647,84 @@ namespace Js
+                END_JS_RUNTIME_CALL(scriptContext);
+                return toStringValue;
+            }
+            return ToStringTagHelper(thisArg, scriptContext, type);
```
We are missing a CrossSite::MarshalVar here, which was triggered by a JSRT debugger test case. Added in the next version.

Yongqu commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +647,84 @@ namespace Js
+            }
+        }
+
+        // Dispatch to new @@toStringTag implemtation.
```
typo :) implementation.

Yongqu commented on lib/Runtime/Library/JavascriptObject.cpp:
```diff
@@ -535,49 +647,84 @@ namespace Js
+        // Dispatch to new @@toStringTag implemtation.
+        if (scriptContext->GetConfig()->IsES6ToStringTagEnabled())
+        {
+            if (type >= TypeIds_TypedArrayMin && type <= TypeIds_TypedArrayMax && !scriptContext->GetThreadContext()->IsScriptActive())
```
hmm is this just for the winrt typed array types? when will this be called from non-scriptactive time?

tcare replied:
This breaks PDM debugger TypedArray tests. I'm not exactly sure why.

tcare commented:
I fixed the crosssite issue, and am leaving the change off by default for now. As discussed offline, external objects need special handling to not break compat before we turn on this change by default. DOM still has some standards work that needs to be done.

I'll follow up in a different PR.

tcare commented:
Ping @akroshg @Yongqu

akroshg commented:
:shipit:
